### ISSUE TITLE
Proposal: Move most documentation to PhpDoc blocks

### ISSUE THREAD

**Issue by franzliedke** [aidantwoods/SecureHeaders#24]
At least for the method documentation, this can be added directly to the code, which can then be used for automatically generating up-to-date documentation, e.g. with https://www.phpdoc.org/.

Example for [`removeHeader()`](https://github.com/aidantwoods/SecureHeaders/wiki/removeHeader):

~~~php
/**
 * Queue a header for removal
 *
 * Upon calling {@see apply}, the header will be removed. This function can
 * be used to prevent automatic headers from being sent.
 *
 * @param string $name Case insensitive name of the header to remove
 * @return void
 */
public function removeHeader($name)
{
    // ...
}
~~~

**Comment by aidantwoods** [comment]
Yes would prefer this 👍

Kind of held off doing it originally when there was only `SecureHeaders.php` (figured it would double the file length). 

Makes a lot more sense to do it now everything is split up though.

Will start going through doing that (though will probably still write some of the docs manually, we'll see how much this adds 😉)

**Comment by aidantwoods** [comment]
So this is probably going to be around the length of most PhpDoc blocks if I were to switch over.

<img width="687" alt="screen shot 2017-02-07 at 09 33 20" src="https://cloud.githubusercontent.com/assets/3288888/22685526/8e76e9aa-ed18-11e6-9d60-c529f0f8d96c.png">

I think that's manageable, and I've thrown together a script to generate the `*.md` files from the "structure.xml" output by PhpDoc, so I can keep the page-per-method structure that I've got in the Wiki currently.

Any best practices RE the extra "example" sections? Or in [this case](https://github.com/aidantwoods/SecureHeaders/wiki/protectedCookie) the "valid constants" section. My guess is that it would be best to write these separately.

**Comment by aidantwoods** [comment]
https://github.com/aidantwoods/SecureHeaders/commit/5844f0acd02b759e1635341bacd1789485b7e188

**Comment by aidantwoods** [comment]
Still have examples to add to that, and have had a bit of fun trying to get `phpdoc` to do what I want it to.

Latest stable comes with an old version of Parsedown (so buggy with lists). Ideally I wanted to turn off markdown parsing (since I want to output .md files, having HTML inside them can cause enclosed markdown not to be rendered). So I ended up with this: https://github.com/aidantwoods/phpDocumentor2/tree/no-markdown, because apparently there's no proper way to turn it off 🤷‍♂️

**Comment by aidantwoods** [comment]
So [this](https://github.com/aidantwoods/MarkdownPhpDocs) should be a not-so-elegant, but "good-enough" script that generates those mentioned `*.md` files 😉

**Comment by franzliedke** [comment]
Nice job, I like it. :+1:

**Comment by aidantwoods** [comment]
It won't work on windows, but meh ;p

I'll probably try do something with example files and [phpdoc's @ example](https://phpdoc.org/docs/latest/references/phpdoc/tags/example.html) support to add in the usage examples/valid constants.

In a couple instances I've made use of nullables in arguments. PHP 7.1 has [actual notation](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.nullable-types) for this, and phpdoc seems to support it (or at least carries the precceding `?` through on the type), so I'll go ahead and add this in to everything that's nullable.

I've modified the output behaviour slightly so that if you write documentation like this

```php
/**
 * @param Type $arg
 * @param ?boolean $optional1 = true
 * @param ?string $optional2 = 'someString'
 */
public function Foo(Type $arg, $optional1 = null, $optional2 = null)
{
    ...
```

Then you'll end up with documentation showing the defined default value after the `=` for those optionals, instead of `null` (which phpdoc shows).

phpdoc's default behaviour is to output `= true` and `= 'someString'` on the first line of the respective descriptions.

**Comment by aidantwoods** [comment]
e.g. for apply, we'll have documentation that looks like this:
<img width="795" alt="screen shot 2017-02-11 at 15 04 41" src="https://cloud.githubusercontent.com/assets/3288888/22854655/8b9a5464-f06b-11e6-8c77-d451fcdb1a59.png">
instead of this
<img width="886" alt="screen shot 2017-02-11 at 15 05 03" src="https://cloud.githubusercontent.com/assets/3288888/22854656/8d5d43c4-f06b-11e6-8465-749c66cd29e1.png">

I'll put a notational remark on the homepage or something to clarify that the `?` means the argument is nullable, and it'll work prior to PHP 7.1.

**Comment by aidantwoods** [comment]
https://github.com/aidantwoods/SecureHeaders/commit/19b90191c189a6e3367016f9268e469431ac004e adds nullable notation

**Comment by aidantwoods** [comment]
Have now added a couple of folders to the [docs folder](https://github.com/aidantwoods/SecureHeaders/tree/2.0/docs) "complements" and "supplements".

"complements" are files that are appended to the docs (to let me separate examples/other sections of docs that phpdoc doesn't accommodate in an easy way).

"supplements" are files that don't exist within (or are to replace) the docs as-interpreted by phpdoc. So this now includes the couple extra pages I had, and also the `->csp` and `->cspro` function docs (which are both polymorphic variadic functions, that are implemented in a version of PHP that doesn't support `...$` – so I can't get phpdoc to make much sense of them).

**Comment by aidantwoods** [comment]
I've also generated the docs and stuck those in folder too (just to give them a quick review next to the original handwritten docs to make sure nothing crucial is missing).

(Just for future reference, they're generated with the following command on the latest version of my [markdown-phpdoc](https://github.com/aidantwoods/MarkdownPhpDocs))

```
markdown-phpdoc -f src/SecureHeaders.php -t docs/generated -c docs/complements/ -s docs/supplements/ -v
```

**Comment by aidantwoods** [comment]
I made some updates to that `markdown-phpdoc` script so that it navigates some of the phpdocumentor internal objects, so that a proper release of phpdocumentor could be used (instead of the hacked fork to just remove the markdown parser).

Unfortunately I can't put it in `composer.json` because of some dependency conflicts. phpdocumentor requires `symfony/console ~2.3` and php-cs-fixer requires `symfony/console ^3.0`.

Since they are both command line tools and don't need to be run in the same php process, they can both be installed separately for running the tests – but unfortunately composer can't facilitate this as far as I can tell.

**Comment by aidantwoods** [comment]
#36 envisions a workaround. Anyone modifying the public SecureHeaders API would have to manually install the script for keeping documentation up to date, or update it manually until phpdocumentor updates that dependency though.

**Comment by aidantwoods** [comment]
Implemented in #36
